### ISSUE TITLE
Consider integrating with Scrapy / utilizing Scrapy to decrease project maintenance burden

### ISSUE THREAD

**Issue by motin** [mozilla/OpenWPM#433]
TL;DR of the first four comments:
Let's base OpenWPM on [Scrapy](https://scrapy.org/) similar to how https://github.com/clemfromspace/scrapy-selenium has it set up. This could decrease the overall project maintenance burden and help with:

* Improve the flexibility of adding new crawl commands
* Support arbitrary link following strategies

<details><summary>Click to show original first posting</summary>
**Original content:**
<p>
I saw this https://github.com/scrapy-plugins/scrapy-splash and https://github.com/clemfromspace/scrapy-selenium. A corresponding `scrapy-openwpm` (or adding openwpm support to splash) could possibly help with:
<ul>
<li>Retrying failed requests - https://github.com/mozilla/OpenWPM/issues/422</li>
<li>Improve logging (by using Scrapy's logging)</li>
<li>Increased resilience (by using Scrapy's battle-tested handling of exceptions and failed requests etc)</li>
<li>Sentry integration for both logging and exceptions - https://github.com/mozilla/OpenWPM/issues/406 since Sentry includes built-in sentry integration</li>
<li>Simplify the readme and barrier-to-entry (since most ordinary scrape/crawl documentation can be referred to upstream)</li>
<li>Minimize the overall project maintenance burden</li>
</p>

<p>I imagine the implementation / integration being most similar to https://github.com/scrapy-plugins/scrapy-splash, with OpenWPM (maybe as a fork of https://github.com/scrapinghub/splash ?) set up as a middleware which pre-"renders" (processes) each request on demand from scrapy and stores all relevant information in S3/locally with an appropriate visit_id. The scrapy spider can then simply save general information about each successful request, including the visit_id, and final analysis of results can then use this information from scrapy to determine which visit_ids are associated with successful requests (aggregated in S3/locally). </p>
</details>

**Comment by motin** [comment]
Splash can be configured to include the adblock extension (https://github.com/TeamHG-Memex/aquarium#options) so configuring with the openwpm extension should not be to much hassle (given compatibilty of the openwpm extension with one of the supported browser engines in splash).

If the openwpm extension in turn uses websockets to send the data, one would only need to fire up a standalone websocket-powered listener / aggregator with splash, and things should work pretty well.

**Comment by motin** [comment]
Note that Splash used to be webkit only, but support for blink (chromium) was merged 11 days ago in https://github.com/scrapinghub/splash/pull/867

**Comment by motin** [comment]
Looking through more of the splash code, a blocker appears: Splash literally only instruments webkit/chromium but includes no browser features. This makes the splash-route a no-go, since we want to emulate browser-based browsing as close as possible. 

A more relevant approach is https://github.com/clemfromspace/scrapy-selenium

**Comment by englehardt** [comment]
This is out of scope.
